README.md
# Aliyun OSS plugin for [Fluentd](http://github.com/fluent/fluentd)

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it whatever you want.

## Overview
**Fluent OSS output plugin** buffers event logs in local files and uploads them to OSS periodically in background threads.

This plugin splits events by using the timestamp of event logs. For example,  a log '2019-04-09 message Hello' is reached, and then another log '2019-04-10 message World' is reached in this order, the former is stored in "20190409.gz" file, and latter in "20190410.gz" file.

**Fluent OSS input plugin** reads data from OSS periodically.

This plugin uses MNS on the same region of the OSS bucket. We must setup MNS and OSS event notification before using this plugin.

[This document](https://help.aliyun.com/document_detail/52656.html) shows how to setup MNS and OSS event notification.

This plugin will poll events from MNS queue and extract object keys from these events, and then will read those objects from OSS.

## Installation

Simply use RubyGems(https://rubygems.org/gems/fluent-plugin-aliyun-oss)!

If you are using td-agent, please make sure to use td-agent’s td-agent-gem command(**`/usr/sbin/td-agent-gem install fluent-plugin-aliyun-oss`**). td-agent has own Ruby so you should install gems into td-agent’s Ruby, not other Ruby. Otherwise (e.g. you use the command belonging to system, rvm, `gem install fluent-plugin-aliyun-oss`, etc.), you won’t be able to find your “installed” plugins.

```bash
[root@master td-agent]# /usr/sbin/td-agent-gem install fluent-plugin-aliyun-oss

Building native extensions.  This could take a while...
Successfully installed unf_ext-0.0.7.6
Fetching: unf-0.1.4.gem (100%)
Successfully installed unf-0.1.4
Fetching: domain_name-0.5.20180417.gem (100%)
Successfully installed domain_name-0.5.20180417
Fetching: http-cookie-1.0.3.gem (100%)
Successfully installed http-cookie-1.0.3
Fetching: rest-client-2.0.2.gem (100%)
Successfully installed rest-client-2.0.2
Fetching: aliyun-sdk-0.7.0.gem (100%)
Building native extensions.  This could take a while...
Successfully installed aliyun-sdk-0.7.0
Fetching: fluent-plugin-aliyun-oss-0.0.1.gem (100%)
Successfully installed fluent-plugin-aliyun-oss-0.0.1
Parsing documentation for unf_ext-0.0.7.6
Installing ri documentation for unf_ext-0.0.7.6
Parsing documentation for unf-0.1.4
Installing ri documentation for unf-0.1.4
Parsing documentation for domain_name-0.5.20180417
Installing ri documentation for domain_name-0.5.20180417
Parsing documentation for http-cookie-1.0.3
Installing ri documentation for http-cookie-1.0.3
Parsing documentation for rest-client-2.0.2
Installing ri documentation for rest-client-2.0.2
Parsing documentation for aliyun-sdk-0.7.0
Installing ri documentation for aliyun-sdk-0.7.0
Parsing documentation for fluent-plugin-aliyun-oss-0.0.1
Installing ri documentation for fluent-plugin-aliyun-oss-0.0.1
Done installing documentation for unf_ext, unf, domain_name, http-cookie, rest-client, aliyun-sdk, fluent-plugin-aliyun-oss after 10 seconds
7 gems installed
```

Then, you can check installed plugin

```bash
[root@master td-agent]# /usr/sbin/td-agent-gem list fluent-plugin-aliyun-oss

*** LOCAL GEMS ***

fluent-plugin-aliyun-oss (0.0.1)
```

## Development

### 1. Plugin Developement and Testing

#### Code
- Install dependencies

```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests
You should set environment variables like below:

test_out_oss.rb
```sh
STORE_AS="" OSS_ENDPOINT="" ACCESS_KEY_ID="" ACCESS_KEY_SECRET="" OSS_BUCKET="" OSS_PATH=""  bundle exec rspec test/plugin/test_out_oss.rb
```

test_in_oss.rb
```sh
STORE_AS=""  OSS_ENDPOINT="" ACCESS_KEY_ID="" ACCESS_KEY_SECRET="" OSS_BUCKET="" MNS_ENDPOINT="" MNS_QUEUE=""  bundle exec rspec test/plugin/test_in_oss.rb

```

## Usage
This is an example of fluent config.

It will read data posted by HTTP and buffer data to local directory before writing to OSS.
You can try it by running curl command:
```bash
[root@master td-agent]# while [[ 1 ]]; do curl -X POST -d 'json={"json":"message"}' http://localhost:8888/debug.test; done
```
	<match debug.*>
	  @type oss
	  endpoint <OSS endpoint to connect to>
	  bucket <Your Bucket>
	  access_key_id <Your Access Key>
	  access_key_secret <Your Secret Key>
	  path fluent-oss/logs
	  auto_create_bucket true
	  key_format %{path}/%{time_slice}_%{index}_%{thread_id}.%{file_extension}
	  store_as gzip
	  <buffer tag,time>
	    @type file
	    path /var/log/fluent/oss
	    timekey 60 # 1 min partition
	    timekey_wait 20s
	    #timekey_use_utc true
	  </buffer>
	  <format>
	    @type json
	  </format>
	</match>
	
	# HTTP input
	# POST http://localhost:8888/<tag>?json=<json>
	# POST http://localhost:8888/td.myapp.login?json={"user"%3A"me"}
	# @see http://docs.fluentd.org/articles/in_http
	<source>
	  @type http
	  @id input_http
	  port 8888
	</source>


## Configuration: Output Plugin
This plugin supports the following configuration options

|Configuration|Type|Required|Comments|Default|
|:---:|:---:|:---:|:---:|:---|
|endpoint|string|Yes|OSS endpoint to connect|
|bucket|string|Yes|Your OSS bucket name|
|access_key_id|string|Yes|Your access key id|
|access_key_secret|string|Yes|Your access secret key|
|path|string|No|Prefix that added to the generated file name|fluent/logs|
|oss_sdk_log_dir|string|No|OSS SDK log directory|/var/log/td-agent|
|upload_crc_enable|bool|No|Enable upload crc check|true|
|download_crc_enable|bool|No|Enable download crc check|true|
|open_timeout|integer|No|Timeout seconds for open connections|10|
|read_timeout|integer|No|Timeout seconds for read response|120|
|key_format|string|No|The format of OSS object keys|%{path}/%{time_slice}\_%{index}\_%{thread_id}.%{file_extension}|
|store_as|string|No|Archive format on OSS|gzip|
|auto_create_bucket|bool|No|Create OSS bucket if it does not exists|true|
|overwrite|bool|No|Overwrite already existing OSS path|false|
|check_bucket|bool|No|Check bucket if exists or not|true|
|check_object|bool|No|Check object before creation|true|
|hex_random_length|integer|No|The length of `%{hex_random}` placeholder(4-16)|4|
|index_format|string|No|`sprintf` format for `%{index}`|%d|
|warn_for_delay|time|No|Set a threshold of events latency and mark these slow events as delayed, output warning logs if delayed events were put into OSS|nil|

### Some configuration details
**key_format**

The format of OSS object keys. You can use the following built-in variables to generate keys dynamically:
*   %{path}
*   %{time_slice}
*   %{index}
*   %{file_extension}
*   %{hex_random}
*   %{uuid_flush}
*   %{thread_id}

* %{path} is exactly the value of **path** configured in the configuration file.
E.g., "fluent/logs" in the example configuration above.
* %{time_slice} is the time-slice in text that are formatted with **time_slice_format**.
* %{index} is the sequential number starts from 0, increments when multiple files are uploaded to OSS in the same time slice.
* %{file_extension} depends on **store_as** parameter.
* %{thread_id} is the unique ids of flush threads(flush thread number is define by `flush_thread_count`). You can use %{thread_id} with other built-in variables to make unique object names. 
* %{uuid_flush} a uuid that is renewed everytime the buffer is flushed. If you want to use this placeholder, install `uuidtools` gem first.
* %{hex_random} a random hex string that is renewed for each buffer chunk, not
guaranteed to be unique. This is used for performance tuning as the article below described,
[OSS performance best practice](https://help.aliyun.com/document_detail/64945.html).
You can configure the length of string with a
`hex_random_length` parameter (Default is 4).

The default format is `%{path}/%{time_slice}_%{index}_%{thread_id}.%{file_extension}`.
For instance, using the example configuration above, actual object keys on OSS
will be something like(flush_thread_count is 1):

    "fluent-oss/logs_20190410-10_15_0_69928273148640.gz"
    "fluent-oss/logs_20190410-10_16_0_69928273148640.gz"
    "fluent-oss/logs_20190410-10_17_0_69928273148640.gz"
    
With the configuration(flush_thread_count is 2):

    key_format %{path}/events/ts=%{time_slice}/events_%{index}_%{thread_id}.%{file_extension}
    time_slice_format %Y%m%d-%H
    path fluent-oss/logs

You get:

    fluent-oss/logs/events/ts=20190410-10/events_0_69997953090220.gz
    fluent-oss/logs/events/ts=20190410-10/events_0_69997953090620.gz
    fluent-oss/logs/events/ts=20190410-10/events_1_69997953090220.gz
    fluent-oss/logs/events/ts=20190410-10/events_1_69997953090620.gz
    fluent-oss/logs/events/ts=20190410-10/events_2_69997953090220.gz
    fluent-oss/logs/events/ts=20190410-10/events_2_69997953090620.gz
    
This plugin also supports add hostname to the final object keys, with the configuration:

**Note:** You should add double quotes to value of `key_format` if use this feature

    key_format "%{path}/events/ts=%{time_slice}/#{Socket.gethostname}/events_%{index}_%{thread_id}.%{file_extension}"
    time_slice_format %Y%m%d-%H
    path fluent-oss/logs
    
You get(flush_thread_count is 1):

    fluent-oss/logs/events/ts=20190410-10/master/events_0_70186087552680.gz
    fluent-oss/logs/events/ts=20190410-10/master/events_1_70186087552680.gz
    fluent-oss/logs/events/ts=20190410-10/master/events_2_70186087552680.gz
    
**store_as**

archive format on OSS. You can use several format:
*   gzip (default)
*   json
*   text
*   lzo (Need lzop command)
*   lzma2 (Need xz command)
*   gzip_command (Need gzip command)
    *   This compressor uses an external gzip command, hence would result in
        utilizing CPU cores well compared with `gzip`
            
**auto_create_bucket**

Create OSS bucket if it does not exists. Default is true.

**check_bucket**

Check configured bucket if it exists or not. Default is true.
When it is false, fluentd will not check the existence of the configured bucket.
This is the case where bucket will be pre-created before running fluentd.

**check_object**

Check object before creation if it exists or not. Default is true.

When it is false, key_format will be %{path}/%{time_slice}\_%{hms_slice}\_%{thread_id}.%{file_extension} by default where,
hms_slice will be time-slice in hhmmss format. With hms_slice and thread_id, each object is unique.
Example object name, assuming it is created on 2019/04/10 10:30:54 AM 20190410_103054_70186087552260.txt (extension can be anything as per user's choice)

**path**

Path prefix of the files on OSS. Default is "fluent-oss/logs".

**time_slice_format**

Format of the time used as the file name. Default is '%Y%m%d'. Use
'%Y%m%d%H' to split files hourly.

**utc**

Use UTC instead of local time.

**hex_random_length**

The length of `%{hex_random}` placeholder. Default is 4.

**index_format**

`%{index}` is formatted by [sprintf](http://ruby-doc.org/core-2.2.0/Kernel.html#method-i-sprintf) using this format_string. Default is '%d'. Zero padding is supported e.g. `%04d` to ensure minimum length four digits. `%{index}` can be in lowercase or uppercase hex using '%x' or '%X'

**overwrite**

Overwrite already existing path. Default is false, which raises an error
if an OSS object of the same path already exists, or increment the
`%{index}` placeholder until finding an absent path.

**warn_for_delay**

Set a threshold to treat events as delay, output warning logs if delayed events were put into OSS.

## Configuration: Input Plugin

|Configuration|Type|Required|Comments|Default|
|:---:|:---:|:---:|:---:|:---|
|endpoint|string|Yes|OSS endpoint to connect|
|bucket|string|Yes|Your OSS bucket name|
|access_key_id|string|Yes|Your access key id|
|access_key_secret|string|Yes|Your access secret key|
|oss_sdk_log_dir|string|No|OSS SDK log directory|/var/log/td-agent|
|upload_crc_enable|bool|No|Enable upload crc check|true|
|download_crc_enable|bool|No|Enable download crc check|true|
|open_timeout|integer|No|Timeout seconds for open connections|10|
|read_timeout|integer|No|Timeout seconds for read response|120|
|store_as|string|No|Archive format on OSS|gzip|
|flush_batch_lines|integer|No|Flush to down streams every `flush_batch_lines` lines.|10000|
|flush_pause_milliseconds|integer|No|Sleep interval between two flushes to downstream.|1|
|store_local|bool|No|Store OSS Objects to local or memory before parsing(Used for objects with `text`/`json`/`gzip` formats)|true|
|mns|configuration section|Yes|MNS configurations|

### Usage
This is an example of fluent config.

    <source>
      @type oss
      endpoint <OSS endpoint to connect to>
      bucket <Your Bucket>
      access_key_id <Your Access Key>
      access_key_secret <Your Secret Key>
      flush_batch_lines 1000
      <mns>
        endpoint <MNS endpoint to connect to, E.g.,{account-id}.mns.cn-zhangjiakou-internal.aliyuncs.com>
        queue <MNS queue>
        wait_seconds 10
        poll_interval_seconds 10
      </mns>
    </source>

### Some configuration details

**store_as**
archive format on OSS. You can use several format:
*   gzip (default)
*   json
*   text
*   lzo (Need lzop command)
*   lzma2 (Need xz command)
*   gzip_command (Need gzip command)
    *   This compressor uses an external gzip command, hence would result in
        utilizing CPU cores well compared with `gzip`

**flush_batch_lines**

Flush to down streams every `flush_batch_lines` lines.

**flush_pause_milliseconds**

Sleep interval between two flushes to downstream. Default is 1ms, and wil not sleep if `flush_pause_milliseconds` is less than or equal to 0.

**store_local(default is true)**

Store OSS Objects to local or memory before parsing(Used for objects with `text`/`json`/`gzip` formats).

Objects with `lzo`/`lzma2`/`gzip_command` formats are always stored to local directory before parsing.

**format**

Parse a line as this format in the OSS object. Supported formats are "apache_error", "apache2", "syslog", "json", "tsv", "ltsv", "csv", "nginx" and "none".

**mns**

[MNS consume messages](https://help.aliyun.com/document_detail/35136.html)

*   endpoint
*   queue
*   wait_seconds
*   poll_interval_seconds  Poll messages interval from MNS

For more details about mns configurations, please view MNS documentation in the link above.

## Website, license, et. al.

| Web site          | http://fluentd.org/                       |
|-------------------|-------------------------------------------|
| Documents         | http://docs.fluentd.org/                  |
| Source repository | http://github.com/aliyun/fluent-plugin-oss |
| Discussion        | http://groups.google.com/group/fluentd    |
| Author            | Jinhu Wu                        |
| License           | Apache License, Version 2.0               |


ChangeLog
Release 0.0.1 - 2019/04/23
  * First release

lib/fluent/plugin/out_oss.rb
require 'fluent/plugin/output'
require 'aliyun/oss'
require 'aliyun/sts'

# This is Fluent OSS Output Plugin
# Usage:
# In order to write output data to OSS, you should add configurations like below
# <match pattern>
#   @type oss
#   endpoint              <OSS endpoint to connect to>
#   bucket                <Your bucket name>
#   access_key_id         <Your access key id>
#   access_key_secret     <Your access secret key>
#   path                  <Path prefix of the files on OSS>
#   key_format            %{path}%{time_slice}_%{index}.%{file_extension}
# if you want to use ${tag} or %Y/%m/%d/ like syntax in path/key_format,
# need to specify tag for ${tag} and time for %Y/%m/%d in <buffer> argument.
#   <buffer tag,time>
#     @type file
#     path /var/log/fluent/oss
#     timekey 3600 # 1 hour partition
#     timekey_wait 10m
#     timekey_use_utc true # use utc
#   </buffer>
#   <format>
#     @type json
#   </format>
# </match>
module Fluent
  # Fluent OSS Plugin
  module Plugin
    # OSSOutput class implementation
    class OSSOutput < Output
      Fluent::Plugin.register_output('oss', self)

      helpers :compat_parameters, :formatter, :inject

      desc 'OSS endpoint to connect to'
      config_param :endpoint, :string
      desc 'Your bucket name'
      config_param :bucket, :string
      desc 'Your access key id'
      config_param :access_key_id, :string
      desc 'Your access secret key'
      config_param :access_key_secret, :string
      desc 'Path prefix of the files on OSS'
      config_param :path, :string, default: 'fluent/logs'
      config_param :upload_crc_enable, :bool, default: true
      config_param :download_crc_enable, :bool, default: true
      desc 'Timeout for open connections'
      config_param :open_timeout, :integer, default: 10
      desc 'Timeout for read response'
      config_param :read_timeout, :integer, default: 120

      desc 'OSS SDK log directory'
      config_param :oss_sdk_log_dir, :string, default: '/var/log/td-agent'

      desc 'The format of OSS object keys'
      config_param :key_format, :string, default: '%{path}/%{time_slice}_%{index}_%{thread_id}.%{file_extension}'
      desc 'Archive format on OSS'
      config_param :store_as, :string, default: 'gzip'
      desc 'Create OSS bucket if it does not exists'
      config_param :auto_create_bucket, :bool, default: false
      desc 'Overwrite already existing path'
      config_param :overwrite, :bool, default: false
      desc 'Check bucket if exists or not'
      config_param :check_bucket, :bool, default: true
      desc 'Check object before creation'
      config_param :check_object, :bool, default: true
      desc 'The length of `%{hex_random}` placeholder(4-16)'
      config_param :hex_random_length, :integer, default: 4
      desc '`sprintf` format for `%{index}`'
      config_param :index_format, :string, default: '%d'
      desc 'Given a threshold to treat events as delay, output warning logs if delayed events were put into OSS'
      config_param :warn_for_delay, :time, default: nil

      DEFAULT_FORMAT_TYPE = 'out_file'.freeze

      config_section :format do
        config_set_default :@type, DEFAULT_FORMAT_TYPE
      end

      config_section :buffer do
        config_set_default :chunk_keys, ['time']
        config_set_default :timekey, (60 * 60 * 24)
      end

      MAX_HEX_RANDOM_LENGTH = 16

      def configure(conf)
        compat_parameters_convert(conf, :buffer, :formatter, :inject)

        super

        raise Fluent::ConfigError, 'Invalid oss endpoint' if @endpoint.nil?

        if @hex_random_length > MAX_HEX_RANDOM_LENGTH
          raise Fluent::ConfigError, 'hex_random_length parameter must be '\
                "less than or equal to #{MAX_HEX_RANDOM_LENGTH}"
        end

        unless @index_format =~ /^%(0\d*)?[dxX]$/
          raise Fluent::ConfigError, 'index_format parameter should follow '\
                '`%[flags][width]type`. `0` is the only supported flag, '\
                'and is mandatory if width is specified. '\
                '`d`, `x` and `X` are supported types'
        end

        begin
          @compressor = COMPRESSOR_REGISTRY.lookup(@store_as).new(buffer_type: @buffer_config[:@type], log: log)
        rescue StandardError => e
          log.warn "'#{@store_as}' not supported. Use 'text' instead: error = #{e.message}"
          @compressor = TextCompressor.new
        end

        @compressor.configure(conf)

        @formatter = formatter_create

        process_key_format

        unless @check_object
          if config.has_key?('key_format')
            log.warn "set 'check_object false' and key_format is "\
                    'specified. Check key_format is unique in each '\
                    'write. If not, existing file will be overwritten.'
          else
            log.warn "set 'check_object false' and key_format is "\
                    'not specified. Use '\
                    "'%{path}/%{time_slice}_%{hms_slice}_%{thread_id}.%{file_extension}' "\
                    'for key_format'
            @key_format = '%{path}/%{time_slice}_%{hms_slice}_%{thread_id}.%{file_extension}'
          end
        end

        @configured_time_slice_format = conf['time_slice_format']
        @values_for_oss_object_chunk = {}
        @time_slice_with_tz = Fluent::Timezone.formatter(
          @timekey_zone,
          @configured_time_slice_format || timekey_to_timeformat(@buffer_config['timekey']))
      end

      def timekey_to_timeformat(timekey)
        case timekey
        when nil          then ''
        when 0...60       then '%Y%m%d-%H_%M_%S' # 60 exclusive
        when 60...3600    then '%Y%m%d-%H_%M'
        when 3600...86400 then '%Y%m%d-%H'
        else                   '%Y%m%d'
        end
      end

      def multi_workers_ready?
        true
      end

      def initialize
        super
        @compressor = nil
        @uuid_flush_enabled = false
      end

      def start
        @oss_sdk_log_dir += '/' unless @oss_sdk_log_dir.end_with?('/')
        Aliyun::Common::Logging.set_log_file(@oss_sdk_log_dir + Aliyun::Common::Logging::DEFAULT_LOG_FILE)
        create_oss_client unless @oss

        ensure_bucket if @check_bucket
        super
      end

      def format(tag, time, record)
        r = inject_values_to_record(tag, time, record)
        @formatter.format(tag, time, r)
      end

      def write(chunk)
        index = 0
        metadata = chunk.metadata
        time_slice = if metadata.timekey.nil?
                       ''.freeze
                     else
                       @time_slice_with_tz.call(metadata.timekey)
                     end

        @values_for_oss_object_chunk[chunk.unique_id] ||= {
          '%{hex_random}' => hex_random(chunk)
        }

        if @check_object
          exist_key = nil
          begin
            values_for_oss_key = {
              '%{path}' => @path,
              '%{thread_id}' => Thread.current.object_id.to_s,
              '%{file_extension}' => @compressor.ext,
              '%{time_slice}' => time_slice,
              '%{index}' => sprintf(@index_format, index)
            }.merge!(@values_for_oss_object_chunk[chunk.unique_id])

            values_for_oss_key['%{uuid_flush}'.freeze] = uuid_random if @uuid_flush_enabled

            key = @key_format.gsub(/%{[^}]+}/) do |matched_key|
              values_for_oss_key.fetch(matched_key, matched_key)
            end
            key = extract_placeholders(key, chunk)
            key = key.gsub(/%{[^}]+}/, values_for_oss_key)

            if (index > 0) && (key == exist_key)
              if @overwrite
                log.warn "#{key} already exists, but will overwrite"
                break
              else
                raise "duplicated path is generated. use %{index} in key_format: path = #{key}"
              end
            end

            index += 1
            exist_key = key
          end while @bucket_handler.object_exists?(key)
        else
          hms_slice = Time.now.utc.strftime('%H%M%S')
          hms_slice = Time.now.strftime('%H%M%S') if @local_time

          values_for_oss_key = {
            '%{path}' => @path,
            '%{thread_id}' => Thread.current.object_id.to_s,
            '%{file_extension}' => @compressor.ext,
            '%{time_slice}' => time_slice,
            '%{hms_slice}' => hms_slice
          }.merge!(@values_for_oss_object_chunk[chunk.unique_id])

          values_for_oss_key['%{uuid_flush}'.freeze] = uuid_random if @uuid_flush_enabled

          key = @key_format.gsub(/%{[^}]+}/) do |matched_key|
            values_for_oss_key.fetch(matched_key, matched_key)
          end
          key = extract_placeholders(key, chunk)
          key = key.gsub(/%{[^}]+}/, values_for_oss_key)
        end

        out_file = Tempfile.new('oss-fluent-')
        out_file.binmode
        begin
          @compressor.compress(chunk, out_file)
          out_file.rewind
          log.info "out_oss: write chunk #{dump_unique_id_hex(chunk.unique_id)} with metadata #{chunk.metadata} to oss://#{@bucket}/#{key}, size #{out_file.size}"

          start = Time.now.to_i
          @bucket_handler.put_object(key, file: out_file, content_type: @compressor.content_type)

          log.debug "out_oss: write oss://#{@bucket}/#{key} used #{Time.now.to_i - start} seconds, size #{out_file.length}"
          @values_for_oss_object_chunk.delete(chunk.unique_id)

          if @warn_for_delay
            if Time.at(chunk.metadata.timekey) < Time.now - @warn_for_delay
              log.warn "out_oss: delayed events were put to oss://#{@bucket}/#{key}"
            end
          end
        ensure
          out_file.close(true) rescue nil
        end
      end

      def create_oss_client
        @oss = Aliyun::OSS::Client.new(
          endpoint: @endpoint,
          access_key_id: @access_key_id,
          access_key_secret: @access_key_secret,
          download_crc_enable: @download_crc_enable,
          upload_crc_enable: @upload_crc_enable,
          open_timeout: @open_timeout,
          read_timeout: @read_timeout
        )
      end

      def process_key_format
        if @key_format.include?('%{uuid_flush}')
          # verify uuidtools
          begin
            require 'uuidtools'
          rescue LoadError
            raise Fluent::ConfigError, 'uuidtools gem not found.'\
                  ' Install uuidtools gem first'
          end

          begin
            uuid_random
          rescue => e
            raise Fluent::ConfigError, "generating uuid doesn't work. "\
                  "Can't use %{uuid_flush} on this environment. #{e}"
          end

          @uuid_flush_enabled = true
        end
      end

      def uuid_random
        ::UUIDTools::UUID.random_create.to_s
      end

      def hex_random(chunk)
        unique_hex = Fluent::UniqueId.hex(chunk.unique_id)
        # unique_hex is like (time_sec, time_usec, rand) => reversing gives more randomness
        unique_hex.reverse!
        unique_hex[0...@hex_random_length]
      end

      def ensure_bucket
        unless @oss.bucket_exist?(@bucket)
          if @auto_create_bucket
            log.info "creating bucket #{@bucket} on #{@endpoint}"
            @oss.create_bucket(@bucket)
          else
            raise "the specified bucket does not exist: bucket = #{@bucket}"
          end
        end

        @bucket_handler = @oss.get_bucket(@bucket)
      end

      # Compression base class.
      class Compressor
        include Fluent::Configurable

        attr_reader :log

        def initialize(opts = {})
          super()
          @buffer_type = opts[:buffer_type]
          @log = opts[:log]
        end

        def configure(conf)
          super
        end

        def ext; end

        def content_type; end

        def compress(chunk, file); end

        private

        def check_command(command, encode = nil)
          require 'open3'

          encode = command if encode.nil?
          begin
            Open3.capture3("#{command} -V")
          rescue Errno::ENOENT
            raise Fluent::ConfigError,
                  "'#{command}' utility must be in PATH for #{encode} compression"
          end
        end
      end

      # Gzip compression.
      class GzipCompressor < Compressor
        def ext
          'gz'.freeze
        end

        def content_type
          'application/x-gzip'.freeze
        end

        def compress(chunk, file)
          out = Zlib::GzipWriter.new(file)
          chunk.write_to(out)
          out.finish
        ensure
          begin
            out.finish
          rescue StandardError
            nil
          end
        end
      end

      # Text output format.
      class TextCompressor < Compressor
        def ext
          'txt'.freeze
        end

        def content_type
          'text/plain'.freeze
        end

        def compress(chunk, file)
          chunk.write_to(file)
        end
      end

      # Json compression.
      class JsonCompressor < TextCompressor
        def ext
          'json'.freeze
        end

        def content_type
          'application/json'.freeze
        end
      end

      COMPRESSOR_REGISTRY = Fluent::Registry.new(:oss_compressor_type,
                                                 'fluent/plugin/oss_compressor_')
      {
        'gzip' => GzipCompressor,
        'json' => JsonCompressor,
        'text' => TextCompressor
      }.each do |name, compressor|
        COMPRESSOR_REGISTRY.register(name, compressor)
      end

      def self.register_compressor(name, compressor)
        COMPRESSOR_REGISTRY.register(name, compressor)
      end
    end
  end
end